### PR TITLE
interfaces/cpu-control: allow to control cpufreq tunables

### PR DESCRIPTION
**Pull request by alexmurray**
This is needed to support snaps like `auto-cpufreq` to use strict confinement as per https://forum.snapcraft.io/t/classic-confinement-request-for-auto-cpufreq/15386/16

Signed-off-by: Alex Murray <alex.murray@canonical.com>

### CODE REVIEW

File: interfaces/builtin/cpu_control.go [hunk 1]
```diff
@@ -35,6 +35,24 @@ const cpuControlConnectedPlugAppArmor = `
...
+/sys/devices/system/cpu/cpufreq/policy*/* r,
+/sys/devices/system/cpu/cpufreq/policy*/energy_performance_preference w,
+/sys/devices/system/cpu/cpufreq/policy*/scaling_governor w,
+/sys/devices/system/cpu/cpufreq/policy*/scaling_max_freq w,
+/sys/devices/system/cpu/cpufreq/policy*/scaling_min_freq w,
+/sys/devices/system/cpu/cpufreq/policy*/scaling_setspeed w,
+/sys/devices/system/cpu/cpufreq/boost w,
+
+# https://www.kernel.org/doc/html/latest/admin-guide/pm/intel_pstate.html#user-space-interface-in-sysfs
+/sys/devices/system/cpu/cpufreq/intel_pstate/* r,
```
Comment (bboozzoo) [inline interfaces/builtin/cpu_control.go]: I think this should be `/sys/devices/system/cpu/intel_pstate/`

Comment (zyga) [reply interfaces/builtin/cpu_control.go]: This makes every file there readable and select files are made writable by the rules below.

Comment (alexmurray) [reply interfaces/builtin/cpu_control.go]: No @bboozzoo is right - I added an extra 'cpufreq' into the path which should be removed - will repush a fixed version soon.

**Comment by zyga** [comment]
The failure is caused by `tests/main/ubuntu-core-snapd` - I am working on a solution.

File: interfaces/builtin/cpu_control.go [hunk 2]
```diff
@@ -35,6 +35,24 @@ const cpuControlConnectedPlugAppArmor = `
 /sys/devices/system/cpu/cpu*/online rw,
 /sys/devices/system/cpu/smt/*       r,
 /sys/devices/system/cpu/smt/control w,
+
+# https://www.kernel.org/doc/html/latest/admin-guide/pm/cpufreq.html#policy-interface-in-sysfs
+/sys/devices/system/cpu/cpufreq/* r,
```
Comment (jdstrand) [inline interfaces/builtin/cpu_control.go]: I wonder if this should be: `/sys/devices/system/cpu/cpufreq/{,**} r,`

File: interfaces/builtin/cpu_control.go [hunk 3]
```diff
@@ -35,6 +35,24 @@ const cpuControlConnectedPlugAppArmor = `
...
+/sys/devices/system/cpu/cpufreq/policy*/* r,
```
Comment (jdstrand) [inline interfaces/builtin/cpu_control.go]: With the above suggestion, this rule can go away.

File: interfaces/builtin/cpu_control.go [hunk 4]
```diff
@@ -35,6 +35,24 @@ const cpuControlConnectedPlugAppArmor = `
...
+/sys/devices/system/cpu/cpufreq/policy*/energy_performance_preference w,
+/sys/devices/system/cpu/cpufreq/policy*/scaling_governor w,
+/sys/devices/system/cpu/cpufreq/policy*/scaling_max_freq w,
+/sys/devices/system/cpu/cpufreq/policy*/scaling_min_freq w,
+/sys/devices/system/cpu/cpufreq/policy*/scaling_setspeed w,
+/sys/devices/system/cpu/cpufreq/boost w,
+
+# https://www.kernel.org/doc/html/latest/admin-guide/pm/intel_pstate.html#user-space-interface-in-sysfs
+/sys/devices/system/cpu/intel_pstate/* r,
```
Comment (jdstrand) [inline interfaces/builtin/cpu_control.go]: Perhaps `/sys/devices/system/cpu/intel_pstate/{,*} r,` instead?

**Comment by zyga** [comment]
This is now fixed by merging https://github.com/snapcore/snapd/pull/8129

**Comment by alexmurray** [comment]
Thanks for the review @jdstrand & @zyga - please see the additional commit which includes @jdstrand's suggested changes - I agree this makes more sense to try and make all attributes readable and then only specific ones writable (and this is more future proof if other subdirectories etc are added with additional attributes).

**Comment by AdnanHodzic** [comment]
Thank you everyone. 

IIRC @alexmurray said these changes will be live in 2.44 release of `snapd`. Does anyone have any indication when this release will go live?

**Pull request merged**